add question, modify conditionals (#6)
GeoScripting-WUR/RPythonBasics: index.Rmd

Buildings[3:] # It is included here
```

As you see, Python interprets the missing number as the first or the last element of the array, depending on it's missing before or after the `:`. R does not, and requires you to enter `1` for the first element and `length(YourArray)` (e.g. `length(Buildings)`) for the last one.

```{block, type="alert alert-success"}

> > **Question 4**: Create a subset of `Buildings` with only `Gaia` and `Aurora`, using index slicing in R and Python. Hint: You may wish to use the functions `c()` for R. To vectorize list in Python, `numpy.array()` would be useful, but creating plain list doesn't need it :)

Reviewer: Nice addition, but grammar needs to be a bit polished :)

```

```{block, type="alert alert-success"}
> **Question 5**: In R and Python, slice `LocMat` to select all values below `6` and above `52`. Hint: You can combine conditions by using `&` (AND) or `|` (OR). Note: In Python, always use parentheses around each condition when combining them — otherwise, operator precedence may cause unexpected behavior.
```

## Conditionals and control flow

> We just saw that we can slice arrays based on conditions. In which `&` and `|` are used for both Python and R. Note that the boolean operators `&` and `|` (for *and* and *or* respectively) are valid in both R and Python, but they mean different things!

Reviewer: ```suggestion
We just saw that we can slice arrays based on conditions. Note that the boolean operators `&` and `|` (for *and* and *or* respectively) are valid in both R and Python, but they mean different things!
```

In R, `&` is a *vectorised* form of `&&`, in other words, it can compare boolean vectors against each other:

```{r}
c(TRUE, TRUE, FALSE, FALSE) & c(TRUE, FALSE, TRUE, FALSE)
```

> This is good in some cases, such as slicing arrays using boolean arrays (like in Question.5, where `&&` doesn't really work), and not so good in other cases, like `if` statements that control program flow.  

Reviewer: ```suggestion
This is good in some cases, such as slicing arrays using boolean arrays (like in Question 5, where `&&` doesn't really work), and not so good in other cases, like `if` statements that control program flow.  This is because `if` requires a single `TRUE` or `FALSE`, not an array.
```

> The conditionals `if` and `else` are used to structure our code, so that part of the code only runs when certain conditions are met. In this case, `if (c(TRUE, FALSE))` is impossible to determine and thus will throw an error. The form `&&` only compares one boolean with another boolean and throws an error otherwise, and thus is suitable. 

Reviewer: ```suggestion
The conditionals `if` and `else` are used to structure our code, so that part of the code only runs when certain conditions are met. In this case, `if (c(TRUE, FALSE))` is impossible to determine and thus will result in an error. The form `&&` only compares one boolean with another boolean, which makes sure you do not accidentally use arrays where a single boolean is needed, and thus is suitable. 
```

Here's an example in R:

```{r}
MyLatitude = 52
GaiaLatitude = WURbuildings[WURbuildings[["name"]] == "Gaia", "y"]
AuroraLatitude = WURbuildings[WURbuildings[["name"]] == "Aurora", "y"]

> if (MyLatitude > GaiaLatitude & MyLatitude > AuroraLatitude) {

Reviewer: Is this more confusing or more informative? Keeping in mind that the Python example is now different from the R example.

Author: Yeah I agree, this is more like a typo I guess... Cause in the elseif it's already `&&` again. Sorry for the mess...again!

}
```

The `&` operator can still work in if statements when both sides are single boolean values. However, it is less strict and may lead to unexpected behavior if vectors are accidentally passed.

In Python, `&` is a *bitwise* comparison, that is, it does not work on booleans, but numbers. 
It just so happens that Python implicitly converts booleans to numbers (`True` to 1 and `False` to 0), so it often works with booleans as well, but it is not intended for that.

> This behavior makes them useful in contexts like Question.5, where we compare an array of values against numeric thresholds and want elementwise boolean logic.

Reviewer: ```suggestion
This behavior makes them useful in contexts like Question 5, where we compare an array of values against numeric thresholds and want element-wise boolean logic.
```

  print("I am in between Gaia and Aurora")
```

```{block, type="alert alert-success"}

> > **Question 6**: Why we don't need parentheses here in this Python code? Hint: it's something to do with operator precedence.

Reviewer: ```suggestion
> **Question 6**: Why don't we need parentheses here in this Python code? Hint: it's something to do with operator precedence.
```
